src/components/Channel/Channel.js

```diff
@@ -1,148 +1,674 @@
-     * */
-    doUpdateMessageRequest: PropTypes.func,
-    /** Disables the channel UI if channel is frozen */
-    disableIfFrozenChannel: PropTypes.bool,
-  };
-
-  static defaultProps = {
-    disableKeyboardCompatibleView: false,
-    disableIfFrozenChannel: true,
-    KeyboardCompatibleView,
-  };
-
-  render() {
-    const { t } = this.props;
-    if (!this.props.channel.cid) {
-      return <Text>{t('Please select a channel first')}</Text>;
+const Channel = (props) => {
+  const {
+    channel,
+    children,
+    disableIfFrozenChannel = true,
+    disableKeyboardCompatibleView = false,
+    emojiData = emojiDataDefault,
+    KeyboardCompatibleView = KeyboardCompatibleViewDefault,
+  } = props;
+
+  const { client, isOnline, logger = () => {} } = useContext(ChatContext);
+  const { t } = useContext(TranslationContext);
+
+  const [editing, setEditing] = useState(null);
+  const [error, setError] = useState(false);
+  const [eventHistory, setEventHistory] = useState({});
+  /**
+   * We save the events in state so that we can display event message
+   * next to the message after which it was received, in MessageList.
+   *
+   * e.g., eventHistory = {
+   *   message_id_1: [
+   *     { ...event_obj_received_after_message_id_1__1 },
+   *     { ...event_obj_received_after_message_id_1__2 },
+   *     { ...event_obj_received_after_message_id_1__3 },
+   *   ],
+   *   message_id_2: [
+   *     { ...event_obj_received_after_message_id_2__1 },
+   *     { ...event_obj_received_after_message_id_2__2 },
+   *     { ...event_obj_received_after_message_id_2__3 },
+   *   ]
+   * }
+   */
+  const [hasMore, setHasMore] = useState(true);
+  const [lastRead, setLastRead] = useState();
+  const [loading, setLoading] = useState(true);
+  const [loadingMore, setLoadingMore] = useState(false);
+  const [members, setMembers] = useState(Immutable({}));
+  const [messages, setMessages] = useState(Immutable([]));
+  const [online, setOnline] = useState(isOnline);
+  const [read, setRead] = useState(Immutable({}));
+  const [thread, setThread] = useState(props.thread);
+  const [threadHasMore, setThreadHasMore] = useState(true);
+  const [threadLoadingMore, setThreadLoadingMore] = useState(false);
+  const [threadMessages, setThreadMessages] = useState(
+    channel.state?.threads?.[props.thread?.id] || [],
+  );
+  const [typing, setTyping] = useState(Immutable({}));
+  const [unmounted, setUnmounted] = useState(false);
+  const [watcherCount, setWatcherCount] = useState();
+  const [watchers, setWatchers] = useState(Immutable({}));
+
+  useEffect(() => {
+    logger('Channel component', 'component mount', {
+      tags: ['lifecycle', 'channel'],
+      props,
+    });
+
+    if (channel) initChannel();
+
+    return () => {
+      logger('Channel component', 'component unmount', {
+        tags: ['lifecycle', 'channel'],
+        props,
+      });
+
+      client.off('connection.recovered', handleEvent);
+      channel.off?.(handleEvent);
```

Comment:
do we need to try/catch either of these so we for sure hit the cancels below?

Reply:
just looked through the off methods in in stream-chat-js and there's only potential to throw an error here if an invalid event key is received as the first argument in a two argument call. the event callback function in simply removed from state, so if the listener isn't valid then nothing happens. maybe we create a constants file for all our event names so we never misspell one?

Reply:
think TS will actually help with that instead of a constants file